web-server/src/content/Service/SystemLogs.tsx

```diff
@@ -16,82 +16,171 @@ import { parseLogLine } from '@/utils/logFormatter';
 
 import { MotionBox } from '../../components/MotionComponents';
 
-export const SystemLogs = ({ serviceName }: { serviceName?: ServiceNames }) => {
+const SystemLogs = memo(({ serviceName }: { serviceName?: ServiceNames }) => {
   const { services, loading, logs } = useSystemLogs({ serviceName });
-  const showScrollDownButton = useBoolState(false);
   const containerRef = useRef<HTMLDivElement>(null);
+  const showScrollDownButton = useBoolState(false);
+  const [searchQuery, setSearchQuery] = useState('');
+  const [debouncedSearchQuery, setDebouncedSearchQuery] = useState('');
+  const [currentMatch, setCurrentMatch] = useState(0);
+  const [totalMatches, setTotalMatches] = useState(0);
+  const [highlightedElements, setHighlightedElements] = useState<HTMLElement[]>([]);
+  const currentHighlightRef = useRef<HTMLElement | null>(null);
+  const isInitialLoad = useRef(true);
+  const isSearchingRef = useRef(false);
+  const currentMatchRef = useRef(0);
+  const totalMatchesRef = useRef(0);
+  const highlightedElementsRef = useRef<HTMLElement[]>([]);
+  const isNavigatingRef = useRef(false);
 
-  const scrollDown = () => {
-    containerRef.current.scrollIntoView({ behavior: 'smooth' });
-  };
-
+  // Debounce search query updates
   useEffect(() => {
-    const observer = new IntersectionObserver(
-      (entries) => {
-        entries.forEach((entry) => {
-          if (entry.isIntersecting) {
-            showScrollDownButton.false();
-          } else {
-            showScrollDownButton.true();
-          }
-        });
-      },
-      {
-        threshold: 0
-      }
-    );
-
-    const containerElement = containerRef.current;
+    const timer = setTimeout(() => {
+      setDebouncedSearchQuery(searchQuery);
```

Comment:
I think the pattern you're using is:

- User input
- debounce 300ms
- Update search query
- wait 300ms
- handle search highlighting

This is something I've seen beginners do, and, when it's laid out for you I'm sure you can see what's wrong here.

Debouncing highlighting on user input is definitely a good idea, but your flow should just be:

- user inputs
- debounce 300ms
- trigger highlighting handler

You also shouldn't maintain two states for search queries, one input by the user, one used by the highlighting handler.

P.S.: This is the kind of stuff that causes folks to fail our frontend interviews, so with this learning, you now have an advantage over them. :) Hopefully you'll come out a lot stronger out of this program. :)